overlord/servicestate/service_control_test.go

```diff
@@ -476,3 +476,71 @@ func (s *serviceControlSuite) TestConflict(c *C) {
 	_, err := snapstate.Remove(st, "test-snap", snap.Revision{}, nil)
 	c.Assert(err, ErrorMatches, `snap "test-snap" has "service-control" change in progress`)
 }
+
+func (s *serviceControlSuite) TestUpdateSnapstateServices(c *C) {
+	var tests = []struct {
+		enable                    []string
+		disable                   []string
+		expectedSnapstateEnabled  []string
+		expectedSnapstateDisabled []string
+		changed                   bool
+	}{
+		{
+			changed: false,
+		},
+		{
+			enable:                   []string{"a"},
+			expectedSnapstateEnabled: []string{"a"},
+			changed:                  true,
+		},
+		// enable again does nothing
+		{
+			enable:                   []string{"a"},
+			expectedSnapstateEnabled: []string{"a"},
+			changed:                  false,
+		},
+		{
+			disable:                   []string{"a"},
+			expectedSnapstateDisabled: []string{"a"},
+			changed:                   true,
+		},
+		{
+			enable:                   []string{"a", "c"},
+			expectedSnapstateEnabled: []string{"a", "c"},
+			changed:                  true,
+		},
+		{
+			disable:                   []string{"b"},
+			expectedSnapstateEnabled:  []string{"a", "c"},
+			expectedSnapstateDisabled: []string{"b"},
+			changed:                   true,
+		},
+		{
+			disable:                   []string{"b", "c"},
+			expectedSnapstateEnabled:  []string{"a"},
+			expectedSnapstateDisabled: []string{"b", "c"},
+			changed:                   true,
+		},
+	}
+
+	snapst := snapstate.SnapState{}
```

Comment:
I think this test should also exercise the logic where snapst has some of the services in the ServicesEnabledByHooks / ServicesDisabledByHooks list too. 

Also this single snapst is shared amongst the test cases below, so isn't there some risk of a test case modifying the snapst and changing the outcome for another test? it seems to me that this snapst should just switch to being declared within the loop.

Reply:
It deliberately uses a single SnapState during execution to accumulate changes and check the scenario where ServicesEnabledByHooks / ServicesDisabledByHooks have services already (per your first remark). I can add a comment about that if you think it helps?

Reply:
Added a comment.

Reply:
ah yes this was totally unclear that you explicitly were testing accumulated changes sorry I don't know why I'm giving such poor reviews here today :-(

Reply:
No worries :), thanks for the review!